InternalPythonModules/GPX_Module/GPX_Parser_Module.py

```diff
@@ -0,0 +1,265 @@
+                            except:                            
+                                pass
+
+                            geoPointList.add(GeoWaypoint.GeoTrackPoint(point.latitude, point.longitude, elevation, 0, 0, 0, dateTime))
+                                                                                                             
+                        try:
+                        # Add the trackpoint using the helper class
+                            geoartifact = geoArtifactHelper.addTrack("Trackpoint", geoPointList)
+                        except Blackboard.BlackboardException as e:
+                            if GPXParserDataSourceIngestModuleFactory.debuglevel: self.log(Level.SEVERE, "GPX: Error using geo artifact helper with blackboard " )
+                        except TskCoreException as e:
+                            if GPXParserDataSourceIngestModuleFactory.debuglevel: self.log(Level.SEVERE, "GPX: Error using geo artifact helper tskcoreexception" )
+                            
+                if GPXParserDataSourceIngestModuleFactory.debuglevel: self.log(Level.INFO, "GPX: WAYPOINTS") 
+                for waypoint in gpx.waypoints:
+                    attributes = ArrayList()
+                    art = file.newArtifact(BlackboardArtifact.ARTIFACT_TYPE.TSK_GPS_BOOKMARK)
+
+                    attributes.add(BlackboardAttribute(BlackboardAttribute.ATTRIBUTE_TYPE.TSK_GEO_LATITUDE.getTypeID(), moduleName, waypoint.latitude))
+                    attributes.add(BlackboardAttribute(BlackboardAttribute.ATTRIBUTE_TYPE.TSK_GEO_LONGITUDE.getTypeID(), moduleName, waypoint.longitude))                    
+                    attributes.add(BlackboardAttribute(BlackboardAttribute.ATTRIBUTE_TYPE.TSK_FLAG.getTypeID(), moduleName, "Waypoint"))
+                    attributes.add(BlackboardAttribute(BlackboardAttribute.ATTRIBUTE_TYPE.TSK_NAME.getTypeID(), moduleName, waypoint.name))
+                    attributes.add(BlackboardAttribute(BlackboardAttribute.ATTRIBUTE_TYPE.TSK_PROG_NAME.getTypeID(), moduleName, "GPXParser"))
+
+                    art.addAttributes(attributes)
+                    
+                    try:
+                    # Post the artifact to blackboard
+                       skCase.getBlackboard().postArtifact(art, moduleName)
+                    except Blackboard.BlackboardException as e:
+                        if GPXParserDataSourceIngestModuleFactory.debuglevel: self.log(Level.SEVERE, "GPX: Error using geo artifact helper with blackboard  for waypoints" )
+
+                if GPXParserDataSourceIngestModuleFactory.debuglevel: self.log(Level.INFO, "GPX: ROUTES")
+                for route in gpx.routes:    
+                    firstTimeThru = 0
+                    startingPoint = list()
+                    endingPoint = list()
+                    for point in route.points:
+                        # If first time in loop only populate starting point
```

Comment:
What is the purpose of this logic? It seems like the logic will break if we have a route that has only 1 point in it, because the end point will not be set and the artifact will not be created?

Reply:
A route needs to have a starting and ending point.  If there is only a starting point and no ending point then it is not a route.  Does that make sense?

Reply:
Oh it definitely makes sense. I'm just thinking that if somehow only one GPS point got saved as a "route" (for whatever reason, bug, etc), it might still be of value to examiners?